File: infra/handler/http/admin_reported.go
package http

import (
	"strings"

	dbmodel "github.com/gabrielmoura/nostr-relay-server/infra/db"
	"github.com/gabrielmoura/nostr-relay-server/internal/db"
	"github.com/gofiber/fiber/v2"
	"github.com/nbd-wtf/go-nostr"
)

func ReportedEvents() fiber.Handler {
	return func(c *fiber.Ctx) error {
		limit := adminLimit(c)
		offset := adminOffset(c)
		filters, err := reportedEventsFiltersFromRequest(c)
		if err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
		}
		if normalized, err := normalizeSearchQuery(filters.Query); err == nil {
			filters.Query = normalized
		}

		reported, total, err := db.DbQueries.GetReportedEvents(c.UserContext(), filters, limit, offset)
		if err != nil {
			return internalServerError(c, err)
		}

		authorKeys := make([]string, 0, len(reported))
		for _, item := range reported {
			if item.TargetPubkey != "" {
				authorKeys = append(authorKeys, item.TargetPubkey)
			}
		}

		authorProfiles, err := db.DbQueries.GetProfilesByPublicKeys(c.UserContext(), authorKeys)
		if err != nil {
			return internalServerError(c, err)
		}

		items := make([]adminReportedEventResponse, 0, len(reported))
		for _, item := range reported {
			var targetEvent *nostr.Event
			targetCreatedAt := int64(0)
			evt, err := db.DbQueries.GetEventByID(c.UserContext(), item.TargetEventID)
			if err == nil {
				targetEvent = evt
				targetCreatedAt = int64(evt.CreatedAt)
			}

			targetPubkey := item.TargetPubkey
			if targetPubkey == "" && targetEvent != nil {
				targetPubkey = targetEvent.PubKey
			}

			profile, ok := authorProfiles[targetPubkey]
			if !ok {
				profile = dbmodel.Profile{PublicKey: targetPubkey, Name: targetPubkey}
			}

			displayName := firstNonEmpty(profile.DisplayName, profile.Name, targetPubkey)
			if displayName == "" {
				displayName = "autor desconhecido"
			}

			items = append(items, adminReportedEventResponse{
				TargetEventID:      item.TargetEventID,
				TargetPubkey:       targetPubkey,
				TargetNevent:       neventFromEventID(item.TargetEventID, targetPubkey),
				TargetCreatedAt:    targetCreatedAt,
				TargetCreatedAtISO: formatUnix(targetCreatedAt),
				TargetAuthor: adminEventAuthor{
					Pubkey:      targetPubkey,
					DisplayName: displayName,
					Picture:     profile.Picture,
					NIP05:       profile.Nip05,
				},
				ReportCount:    item.ReportCount,
				LastReported:   item.LastReported,
				LastReportedAt: formatUnix(item.LastReported),
				ReportTypes:    item.ReportTypes,
				TargetEvent:    targetEvent,
			})
		}

		return c.JSON(newAdminPage(items, int(total), limit, offset))
	}
}

func ReportedEventsSummary() fiber.Handler {
	return func(c *fiber.Ctx) error {
		filters, err := reportedEventsFiltersFromRequest(c)
		if err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
		}
		if normalized, err := normalizeSearchQuery(filters.Query); err == nil {
			filters.Query = normalized
		}

		summary, err := db.DbQueries.GetReportedEventsSummary(c.UserContext(), filters)
		if err != nil {
			return internalServerError(c, err)
		}

		authorKeys := make([]string, 0, len(summary.TopAuthors))
		for _, author := range summary.TopAuthors {
			if author.Pubkey != "" {
				authorKeys = append(authorKeys, author.Pubkey)
			}
		}

		authorProfiles, err := db.DbQueries.GetProfilesByPublicKeys(c.UserContext(), authorKeys)
		if err != nil {
			return internalServerError(c, err)
		}

		response := adminReportedEventsSummaryResponse{
			TotalEvents:         summary.TotalEvents,
			TotalReports:        summary.TotalReports,
			UniqueTargetAuthors: summary.UniqueTargetAuthors,
			Timeline:            make([]adminReportedTimelinePointResponse, 0, len(summary.Timeline)),
			ReportTypes:         make([]adminReportedTypeCountResponse, 0, len(summary.ReportTypes)),
			TopAuthors:          make([]adminReportedAuthorCountResponse, 0, len(summary.TopAuthors)),
			TopTargets:          make([]adminReportedTargetCountResponse, 0, len(summary.TopTargets)),
		}

		for _, item := range summary.Timeline {
			response.Timeline = append(response.Timeline, adminReportedTimelinePointResponse{Bucket: item.Bucket, Count: item.Count})
		}
		for _, item := range summary.ReportTypes {
			response.ReportTypes = append(response.ReportTypes, adminReportedTypeCountResponse{Name: item.Name, Count: item.Count})
		}
		for _, item := range summary.TopAuthors {
			displayName := item.Pubkey
			if profile, ok := authorProfiles[item.Pubkey]; ok {
				displayName = firstNonEmpty(profile.DisplayName, profile.Name, item.Pubkey)
			}
			response.TopAuthors = append(response.TopAuthors, adminReportedAuthorCountResponse{Pubkey: item.Pubkey, DisplayName: displayName, Count: item.Count})
		}
		for _, item := range summary.TopTargets {
			response.TopTargets = append(response.TopTargets, adminReportedTargetCountResponse{TargetEventID: item.TargetEventID, Count: item.Count})
		}

		return c.JSON(response)
	}
}

func reportedEventsFiltersFromRequest(c *fiber.Ctx) (dbmodel.ReportedEventsFilters, error) {
	filters := dbmodel.ReportedEventsFilters{
		Query:         strings.TrimSpace(c.Query("q")),
		ReportType:    strings.TrimSpace(c.Query("type")),
		TargetPubkey:  strings.TrimSpace(c.Query("target_pubkey")),
		TargetEventID: strings.TrimSpace(c.Query("target_event_id")),
		Since:         int64(c.QueryInt("since", 0)),
		Until:         int64(c.QueryInt("until", 0)),
	}

	if filters.TargetPubkey != "" {
		normalized, err := normalizePublicKey(filters.TargetPubkey)
		if err != nil {
			return dbmodel.ReportedEventsFilters{}, err
		}
		filters.TargetPubkey = normalized
	}
	if filters.TargetEventID != "" {
		normalized, err := normalizeEventID(filters.TargetEventID)
		if err != nil {
			return dbmodel.ReportedEventsFilters{}, err
		}
		filters.TargetEventID = normalized
	}

	return filters, nil
}

File: infra/dash/src/services/admin-reported.ts
import { adminApolloClient } from "@/graphql/client"
import { EventReportsDocument, ReportedEventsDocument, ReportedEventsSummaryDocument } from "@/graphql/generated/operations"
import { GraphQLApiError } from "@/graphql/helpers"
import type { AdminPage, EventReportsResponse, EventRecord, ReportedEventItem, ReportedEventsSummary } from "@/types/admin"

import { ApiError } from "./admin"

type PageParams = {
  limit: number
  offset: number
}

type ReportedEventsQueryFilters = {
  query: string
  type: string
  target_pubkey?: string
  target_event_id?: string
  since?: number
  until?: number
}

function toGraphApiError(error: unknown): never {
  if (error instanceof GraphQLApiError) {
    throw new ApiError(error.message, error.status, error.details, error.requestId)
  }
  throw error
}

function mapEventRecord(item: any): EventRecord {
  return {
    id: item.id,
    pubkey: item.pubkey,
    created_at: item.createdAt,
    kind: item.kind,
    content: item.content,
    sig: item.sig,
    tags: (item.tags ?? []).map((tag: any) => tag.values),
  }
}

function mapReportedEventItem(item: any): ReportedEventItem {
  const targetAuthor = item.targetAuthor
  const targetPubkey = item.targetPubkey ? String(item.targetPubkey) : targetAuthor?.pubkey ? String(targetAuthor.pubkey) : ""

  return {
    target_event_id: item.targetEventId,
    target_pubkey: targetPubkey || undefined,
    target_nevent: item.targetNevent ?? undefined,
    target_created_at: item.targetCreatedAt ?? undefined,
    target_created_at_iso: item.targetCreatedAtIso ?? undefined,
    target_author: targetAuthor ? {
      pubkey: targetAuthor.pubkey ? String(targetAuthor.pubkey) : targetPubkey,
      display_name: targetAuthor.displayName ? String(targetAuthor.displayName) : targetPubkey || "autor desconhecido",
      picture: targetAuthor.picture ? String(targetAuthor.picture) : undefined,
      nip05: targetAuthor.nip05 ? String(targetAuthor.nip05) : undefined,
    } : undefined,
    report_count: item.reportCount,
    last_reported: item.lastReported,
    last_reported_at: item.lastReportedAt ?? undefined,
    report_types: item.reportTypes ?? [],
    target_event: item.targetEvent ? mapEventRecord(item.targetEvent) : undefined,
  }
}

function buildReportedEventsFilter(filters: ReportedEventsQueryFilters) {
  return {
    q: filters.query || null,
    types: filters.type && filters.type !== "all" ? [filters.type] : null,
  }
}

export async function getEventReports(eventID: string, params: PageParams) {
  try {
    const result = await adminApolloClient.query({
      query: EventReportsDocument,
      variables: { id: eventID, page: { limit: params.limit, offset: params.offset } },
    })
    const data = result.data
    if (!data) {
      throw new ApiError("GraphQL query returned no data")
    }

    return {
      items: data.eventReports.items.map((item: any) => ({
        report_event_id: item.reportEventId,
        reporter_pubkey: item.reporterPubkey,
        reporter_npub: item.reporterNpub ?? undefined,
        reporter_display_name: item.reporterDisplayName,
        reporter_picture: item.reporterPicture ?? undefined,
        reported_event_id: item.reportedEventId ?? undefined,
        reported_pubkey: item.reportedPubkey ?? undefined,
        report_type: item.reportType ?? undefined,
        content: item.content ?? undefined,
        created_at: item.createdAt,
      })),
      total: data.eventReports.pageInfo.total,
    } satisfies EventReportsResponse
  } catch (error) {
    toGraphApiError(error)
  }
}

export async function getReportedEventsPage(filters: ReportedEventsQueryFilters, params: PageParams) {
  try {
    const result = await adminApolloClient.query({
      query: ReportedEventsDocument,
      variables: {
        filter: buildReportedEventsFilter(filters),
        page: { limit: params.limit, offset: params.offset },
      },
    })
    const data = result.data
    if (!data) {
      throw new ApiError("GraphQL query returned no data")
    }

    return {
      items: data.reportedEvents.items.map(mapReportedEventItem),
      total: data.reportedEvents.pageInfo.total,
      limit: data.reportedEvents.pageInfo.limit,
      offset: data.reportedEvents.pageInfo.offset,
      has_more: data.reportedEvents.pageInfo.hasMore,
    } satisfies AdminPage<ReportedEventItem>
  } catch (error) {
    toGraphApiError(error)
  }
}

export async function getReportedEventsSummary(filters: ReportedEventsQueryFilters) {
  try {
    const result = await adminApolloClient.query({
      query: ReportedEventsSummaryDocument,
      variables: {
        filter: buildReportedEventsFilter(filters),
      },
    })
    const data = result.data
    if (!data) {
      throw new ApiError("GraphQL query returned no data")
    }

    return {
      total_events: data.reportedEventsSummary.totalEvents,
      total_reports: data.reportedEventsSummary.totalReports,
      unique_target_authors: data.reportedEventsSummary.uniqueTargetAuthors,
      timeline: data.reportedEventsSummary.timeline.map((item: any) => ({ bucket: item.name, count: item.count })),
      report_types: data.reportedEventsSummary.reportTypes.map((item: any) => ({ name: item.name, count: item.count })),
      top_authors: data.reportedEventsSummary.topAuthors.map((item: any) => ({ pubkey: item.pubkey, display_name: item.displayName ?? item.pubkey, count: item.count })),
      top_targets: data.reportedEventsSummary.topTargets.map((item: any) => ({ target_event_id: item.targetEventId, count: item.count })),
    } satisfies ReportedEventsSummary
  } catch (error) {
    toGraphApiError(error)
  }
}
